test/config-luatest/roles_test.lua

```diff
@@ -0,0 +1,443 @@
+        t.assert_equals(roles_cfg['one'], 1)
+        t.assert_equals(_G.foo, 42 * 2 - 100)
+    end
+
+    helpers.reload_success_case(g, {
+        roles = {one = one},
+        roles_2 = {one = one},
+        options = {
+            ['roles_cfg'] = {one = 2},
+            ['roles'] = {'one'}
+        },
+        options_2 = {
+            ['roles_cfg'] = {one = 1},
+        },
+        verify = verify,
+        verify_2 = verify_2,
+    })
+end
+
+-- Make sure an error is raised if validate_config() does not return true.
+g.test_role_wrong_cfg_error = function(g)
+    local one = [[
+        local function validate_config(cfg)
+        end
+
+        return {
+            init = function() end,
+            validate_config = validate_config,
+            apply_config = function() end,
+            stop = function() end,
+        }
+    ]]
+
+    helpers.failure_case(g, {
+        roles = {one = one},
+        options = {
+            ['roles_cfg'] = {one = 1},
+            ['roles'] = {'one'}
+        },
+        exp_err = 'Wrong config for role one'
+    })
+end
+
+-- Make sure an error is raised if not all methods are present.
+g.test_role_no_method_error = function(g)
+    local one = [[
+        return {
+            validate_config = function() end,
+            apply_config = function() end,
+            stop = function() end,
+        }
+    ]]
+    helpers.failure_case(g, {
+        roles = {one = one},
+        options = {
+            ['roles_cfg'] = {one = 1},
+            ['roles'] = {'one'}
+        },
+        exp_err = 'Role one does not contain method init'
+    })
+
+    one = [[
+        return {
+            init = function() end,
+            apply_config = function() end,
+            stop = function() end,
+        }
+    ]]
+    helpers.failure_case(g, {
+        roles = {one = one},
+        options = {
+            ['roles_cfg'] = {one = 1},
+            ['roles'] = {'one'}
+        },
+        exp_err = 'Role one does not contain method validate_config'
+    })
+
+    one = [[
+        return {
+            init = function() end,
+            validate_config = function() end,
+            stop = function() end,
+        }
+    ]]
+    helpers.failure_case(g, {
+        roles = {one = one},
+        options = {
+            ['roles_cfg'] = {one = 1},
+            ['roles'] = {'one'}
+        },
+        exp_err = 'Role one does not contain method apply_config'
+    })
+
+    one = [[
+        return {
+            init = function() end,
+            validate_config = function() end,
+            apply_config = function() end,
+        }
+    ]]
+    helpers.failure_case(g, {
+        roles = {one = one},
+        options = {
+            ['roles_cfg'] = {one = 1},
+            ['roles'] = {'one'}
+        },
+        exp_err = 'Role one does not contain method stop'
+    })
+end
+
+-- Verify that the error is raised by config:reload() and the same
+-- error appears in the alerts.
```

Comment:
I appreciate your work on the test cases. Thank you!

I would add cases, when `apply_config` fails and when `stop` fails.

Reply:
Done.